From 2a6d1698a7c29f7c01702864bd883a881af8b10c Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Mon, 15 Feb 2021 15:07:25 +0200
Subject: [PATCH] [SW-20154] move gaudi2_arc to gaudi2_arc_gnu

As part of migration of toolchain so that all
embedded projects (arcboot, ppboot and preboot)
will use the same one, this patch moves the files
from gaudi2_arc to gaudi2_arc_gnu

Change-Id: I8d9a8fc9bf9b9ef6837a4bd3bb759455cfcf2871
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/gaudi2_arc/Makefile b/gaudi2_arc/Makefile
deleted file mode 100644
index e69de29..0000000
--- a/gaudi2_arc/Makefile
+++ /dev/null
diff --git a/gaudi2_arc/arcboot_ram.lds b/gaudi2_arc/arcboot_ram.lds
deleted file mode 100644
index 056dca1..0000000
--- a/gaudi2_arc/arcboot_ram.lds
+++ /dev/null
@@ -1,88 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
-OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
-OUTPUT_ARCH(arc)
-
-MEMORY
-{
-	ICCM	: ORIGIN = 0x00000000, LENGTH = 128K
-	SRAM	: ORIGIN = 0x11000000, LENGTH = 48M
-	STM	: ORIGIN = 0x24000000, LENGTH = 32M
-	SPI	: ORIGIN = 0x26000000, LENGTH = 16M
-	SBROM	: ORIGIN = 0x27FC0000, LENGTH = 64K
-	FWSRAM	: ORIGIN = 0x27FD0000, LENGTH = 64K
-	CS	: ORIGIN = 0x2C000000, LENGTH = 32M
-	DCCM	: ORIGIN = 0x80000000, LENGTH = 32K
-}
-
-REGION_ALIAS("REGION_TEXT",	ICCM)
-REGION_ALIAS("REGION_DATA",	DCCM)
-REGION_ALIAS("REGION_SDATA",	DCCM)
-REGION_ALIAS("REGION_SBSS",	DCCM)
-REGION_ALIAS("REGION_BSS",	DCCM)
-REGION_ALIAS("REGION_HEAP",	DCCM)
-REGION_ALIAS("REGION_STACK",	DCCM)
-
-ENTRY(__start)
-
-SECTIONS
-{
-	.text : {
-		app/crt0_ram.o (.text*)
-		*(.text .text.*)
-		_etext = . ;
-	} > REGION_TEXT
-
-	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
-		_data = .;
-		*(.data)
-	} > REGION_DATA
-
-	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
-		__SDATA_BEGIN__ = .;
-		*(.sdata .sdata.* .gnu.linkonce.s.*)
-		_edata = .;
-		PROVIDE (edata = .);
-	} > REGION_SDATA
-
-	.sbss : {
-		__bss_start = .;
-		PROVIDE (__sbss_start = .);
-		PROVIDE (___sbss_start = .);
-		*(.dynsbss)
-		*(.sbss .sbss.* .gnu.linkonce.sb.*)
-		*(.scommon)
-		PROVIDE (__sbss_end = .);
-		PROVIDE (___sbss_end = .);
-	} > REGION_SBSS
-
-	.bss : {
-		*(.dynbss)
-		*(.bss .bss.* .gnu.linkonce.b.*)
-		*(COMMON)
-		/* Align here to ensure that the .bss section occupies space up to
-		_end. Align after .bss to ensure correct alignment even if the
-		.bss section disappears because there are no input sections. */
-		. = ALIGN(32 / 8);
-		_end = .;
-		PROVIDE (end = .);
-	} > REGION_BSS
-
-	.heap : {
-		. = ALIGN(32 / 8);
-		__start_heap = . ;
-		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 1k)) ;
-		__end_heap = . ;
-		. = ALIGN(0x8);
-	} > REGION_HEAP
-
-	.stack : {
-		. = ALIGN(32 / 8);
-		__stack = . ;
-		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 512)) ;
-		__stack_top = . ;
-	} > REGION_STACK
-}
diff --git a/gaudi2_arc/arcboot_rom.lds b/gaudi2_arc/arcboot_rom.lds
deleted file mode 100644
index 1721eb2..0000000
--- a/gaudi2_arc/arcboot_rom.lds
+++ /dev/null
@@ -1,96 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
-OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
-OUTPUT_ARCH(arc)
-
-MEMORY
-{
-	ICCM	: ORIGIN = 0x00000000, LENGTH = 128K
-	SRAM	: ORIGIN = 0x11000000, LENGTH = 48M
-	STM	: ORIGIN = 0x24000000, LENGTH = 32M
-	SPI	: ORIGIN = 0x26000000, LENGTH = 16M
-	SBROM	: ORIGIN = 0x27FC0000, LENGTH = 64K
-	FWSRAM	: ORIGIN = 0x27FD0000, LENGTH = 64K
-	CS	: ORIGIN = 0x2C000000, LENGTH = 32M
-	DCCM	: ORIGIN = 0x80000000, LENGTH = 32K
-}
-
-REGION_ALIAS("REGION_IVT",	SBROM)
-REGION_ALIAS("REGION_TEXT",	SBROM)
-REGION_ALIAS("REGION_DATA",	DCCM)
-REGION_ALIAS("REGION_SDATA",	DCCM)
-REGION_ALIAS("REGION_SBSS",	DCCM)
-REGION_ALIAS("REGION_BSS",	DCCM)
-REGION_ALIAS("REGION_HEAP",	DCCM)
-REGION_ALIAS("REGION_STACK",	DCCM)
-
-ENTRY(__start)
-
-SECTIONS
-{
-	_dccm_begin = ORIGIN(DCCM);
-	_dccm_end = ORIGIN(DCCM) + LENGTH(DCCM);
-
-	.ivt : {
-		KEEP (*(.ivt));
-	} > REGION_IVT
-
-	.text : {
-		app/crt0_rom.o (.text*)
-		*(.text .text.*)
-		_etext = . ;
-	} > REGION_TEXT
-
-	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
-		_data = .;
-		*(.data)
-	} > REGION_DATA
-
-	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
-		__SDATA_BEGIN__ = .;
-		*(.sdata .sdata.* .gnu.linkonce.s.*)
-		_edata = .;
-		PROVIDE (edata = .);
-	} > REGION_SDATA
-
-	.sbss : {
-		__bss_start = .;
-		PROVIDE (__sbss_start = .);
-		PROVIDE (___sbss_start = .);
-		*(.dynsbss)
-		*(.sbss .sbss.* .gnu.linkonce.sb.*)
-		*(.scommon)
-		PROVIDE (__sbss_end = .);
-		PROVIDE (___sbss_end = .);
-	} > REGION_SBSS
-
-	.bss : {
-		*(.dynbss)
-		*(.bss .bss.* .gnu.linkonce.b.*)
-		*(COMMON)
-		/* Align here to ensure that the .bss section occupies space up to
-		_end. Align after .bss to ensure correct alignment even if the
-		.bss section disappears because there are no input sections. */
-		. = ALIGN(32 / 8);
-		_end = .;
-		PROVIDE (end = .);
-	} > REGION_BSS
-
-	.heap : {
-		. = ALIGN(32 / 8);
-		__start_heap = . ;
-		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 5k)) ;
-		__end_heap = . ;
-		. = ALIGN(0x8);
-	} > REGION_HEAP
-
-	.stack : {
-		. = ALIGN(32 / 8);
-		__stack = . ;
-		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 2k)) ;
-		__stack_top = . ;
-	} > REGION_STACK
-}
diff --git a/gaudi2_arc/asic.c b/gaudi2_arc/asic.c
deleted file mode 100644
index 7b6e440..0000000
--- a/gaudi2_arc/asic.c
+++ /dev/null
@@ -1,4 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
diff --git a/gaudi2_arc/asic_arcboot_ram.lds b/gaudi2_arc/asic_arcboot_ram.lds
deleted file mode 100644
index 827fe1f..0000000
--- a/gaudi2_arc/asic_arcboot_ram.lds
+++ /dev/null
@@ -1,6 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
- INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc/asic_arcboot_rom.lds b/gaudi2_arc/asic_arcboot_rom.lds
deleted file mode 100644
index 9e1e858..0000000
--- a/gaudi2_arc/asic_arcboot_rom.lds
+++ /dev/null
@@ -1,6 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
- INCLUDE arcboot_rom.lds
diff --git a/gaudi2_arc/common.c b/gaudi2_arc/common.c
deleted file mode 100644
index af86bf7..0000000
--- a/gaudi2_arc/common.c
+++ /dev/null
@@ -1,66 +0,0 @@
-// SPDX-License-Identifier:     GPL-2.0+
-/*
- * Copyright (C) 2021 HabanaLabs Ltd.
- * All Rights Reserved.
- *
- */
-
-#include <unistd.h>
-#include <efuse.h>
-#include <pcie.h>
-#include <arcboot.h>
-#include <psoc_efuse_regs.h>
-#include <psoc_global_conf_masks.h>
-#include <psoc_reset_conf_masks.h>
-#include <psoc_reset_conf_regs.h>
-
-#define ARCBOOT_STATUS_REG_SZ_BITS	32
-
-void arcboot_rom_status_update(enum arcboot_rom_progress progress)
-{
-	uint32_t reg, offset;
-
-	offset = progress / ARCBOOT_STATUS_REG_SZ_BITS;
-	progress = progress % ARCBOOT_STATUS_REG_SZ_BITS;
-
-	reg = regmap_readl((mmPSOC_SECURITY_SW_STATUS_0 + offset * 4));
-	reg |= BIT(progress);
-	regmap_writel(reg, (mmPSOC_SECURITY_SW_STATUS_0 + offset  * 4));
-}
-
-uint32_t get_efuse_read_reg(void)
-{
-	return mmPSOC_EFUSE_READ_CTRL;
-}
-
-uint32_t get_mem_cfg_mask(void)
-{
-	return PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_MEM_SIZE_MASK;
-}
-
-uint32_t get_fld_size_shift(void)
-{
-	return PSOC_GLOBAL_CONF_SPI_DMA_CTRL_MEM_SIZE_SHIFT;
-}
-
-void update_warm_reset_config(void)
-{
-	regmap_writel(PSOC_RESET_CONF_PSOC_PRSTN_RST_CFG_EN_MASK,
-			mmPSOC_RESET_CONF_PSOC_PRSTN_RST_CFG);
-	regmap_writel(PSOC_RESET_CONF_ARC_PRSTN_RST_CFG_EN_MASK,
-			mmPSOC_RESET_CONF_ARC_PRSTN_RST_CFG);
-}
-
-uint8_t get_sris_mode_from_strap()
-{
-	uint32_t boot_strap;
-
-	boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H);
-
-	return !!(boot_strap & PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H_SRIS_MODE_MASK);
-}
-
-void enable_flr_handling_fsm(void)
-{
-	/* TODO: No HW FSM bits in Gaudi2, wait for ORON to comeback */
-}
diff --git a/gaudi2_arc/dev_security.h b/gaudi2_arc/dev_security.h
deleted file mode 100644
index 28a618e..0000000
--- a/gaudi2_arc/dev_security.h
+++ /dev/null
@@ -1,38 +0,0 @@
-// SPDX-License-Identifier:     GPL-2.0+
-/*
- * Copyright (C) 2021 HabanaLabs Ltd.
- * All Rights Reserved.
- *
- */
-
-#ifndef _GAUDI2_DEV_SECURITY_H_
-#define _GAUDI2_DEV_SECURITY_H_
-
-#include <sha512.h>
-
-#define NUM_KEYS_IN_ROM					5
-#define NUM_KEYS_IN_EFUSE				3
-#define SECURITY_DWORD_PER_KEY				12
-#define SECURITY_KEY_HASH_COUNT				3
-
-enum security_key_location {
-	KEY_LOCATION_EFUSE,
-	KEY_LOCATION_ROM
-};
-
-/* size of key properties in bytes */
-#define KEY_PROP_RR_SZ  				384
-#define KEY_PROP_MOD_SZ					384
-#define KEY_PROP_EXP_SZ					8
-#define KEY_PROP_INV_SZ					4
-#define KEY_PROP_NB_SZ					4
-#define KEY_PROP_HDR_SZ					8
-
-/* 384 + 384 + 8 + 4 + 4 + 8 = 792 */
-#define TOTAL_PUB_KEY_PROPS_SZ  (KEY_PROP_RR_SZ + KEY_PROP_MOD_SZ + \
-				KEY_PROP_EXP_SZ + KEY_PROP_INV_SZ + \
-				KEY_PROP_NB_SZ + KEY_PROP_HDR_SZ)
-
-extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
-
-#endif /* _GAUDI2_DEV_SECURITY_H_ */
diff --git a/gaudi2_arc/onix.c b/gaudi2_arc/onix.c
deleted file mode 100644
index f82fc69..0000000
--- a/gaudi2_arc/onix.c
+++ /dev/null
@@ -1,29 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
-#include <unistd.h>
-#include "arcboot.h"
-#include "efuse.h"
-#include <common.h>
-#include <psoc_global_conf_masks.h>
-
-#define EFUSE_SYSCFG_ONIX_DEFAULT \
-				(EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI | \
-				 EFUSE_SYSCFG_KEY_LOCATION_SPI | \
-				 EFUSE_SYSCFG_REVOKATION_LOCATION_SPI | \
-				 EFUSE_SYSCFG_ENABLE_VALIDATION_MASK | \
-				 EFUSE_SYSCFG_GPIO02_RECOVERY_CONTROL)
-
-int fuse_read(uint32_t word, uint32_t *val)
-{
-	int ret = -1;
-
-	if (word == EFUSE_SYSCFG_OFFSET) {
-		*val = EFUSE_SYSCFG_ONIX_DEFAULT;
-		ret = 0;
-	}
-
-	return ret;
-}
diff --git a/gaudi2_arc/onix_arcboot_ram.lds b/gaudi2_arc/onix_arcboot_ram.lds
deleted file mode 100644
index 827fe1f..0000000
--- a/gaudi2_arc/onix_arcboot_ram.lds
+++ /dev/null
@@ -1,6 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
- INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc/onix_arcboot_rom.lds b/gaudi2_arc/onix_arcboot_rom.lds
deleted file mode 100644
index 1a9a352..0000000
--- a/gaudi2_arc/onix_arcboot_rom.lds
+++ /dev/null
@@ -1,12 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
-INCLUDE arcboot_rom.lds
-
-/* This is SRAM address, used for Onix targets */
-MEMORY
-{
-	SBROM	: ORIGIN = 0x11000000, LENGTH = 64K
-}
diff --git a/gaudi2_arc/pldm.c b/gaudi2_arc/pldm.c
deleted file mode 100644
index 7b6e440..0000000
--- a/gaudi2_arc/pldm.c
+++ /dev/null
@@ -1,4 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
diff --git a/gaudi2_arc/pldm_arcboot_ram.lds b/gaudi2_arc/pldm_arcboot_ram.lds
deleted file mode 100644
index 827fe1f..0000000
--- a/gaudi2_arc/pldm_arcboot_ram.lds
+++ /dev/null
@@ -1,6 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
- INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc/pldm_arcboot_rom.lds b/gaudi2_arc/pldm_arcboot_rom.lds
deleted file mode 100644
index 9e1e858..0000000
--- a/gaudi2_arc/pldm_arcboot_rom.lds
+++ /dev/null
@@ -1,6 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
- INCLUDE arcboot_rom.lds
diff --git a/gaudi2_arc/secureboot.c b/gaudi2_arc/secureboot.c
deleted file mode 100644
index 2ef2f58..0000000
--- a/gaudi2_arc/secureboot.c
+++ /dev/null
@@ -1,196 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
-#include <string.h>
-#include <arcboot.h>
-#include <efuse.h>
-#include <secureboot.h>
-#include <security.h>
-#include <dev_security.h>
-#include <rsa.h>
-#include <sha512.h>
-
-
-const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN] = {
-	/* Habana Public Key */
-	{
-		0x09, 0xa8, 0xdd, 0xac, 0xbe, 0x8a, 0x6d, 0x34,
-		0x94, 0x23, 0x62, 0x7c, 0xd6, 0x11, 0xe8, 0x15,
-		0xb4, 0x44, 0xe2, 0x8d, 0x81, 0xc9, 0xc5, 0xd0,
-		0x49, 0x3d, 0xd6, 0x4f, 0x10, 0xac, 0xdf, 0x83,
-		0x04, 0x61, 0xb5, 0xde, 0x3e, 0x58, 0x5a, 0x73,
-		0x18, 0xb4, 0x56, 0xbc, 0xcc, 0x91, 0xf0, 0xed
-	},
-	/* Additional Keys - TODO */
-	{
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0
-	},
-	{
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0
-	},
-	{
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0
-	},
-	{
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
-		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0
-	}
-};
-
-static int security_check_key_revoke_status(uint32_t key_index)
-{
-	uint8_t max_key_index = NUM_KEYS_IN_EFUSE, key_revoke;
-
-	/*
-	 * Key index ranges from:
-	 * 0 - 4: if key location == ROM (1)
-	 * 0 - 2: if key location == eFUSE (0)
-	 */
-	/* check efuse for key location */
-	if ((arcboot_rom_global.config_fuse >> EFUSE_SYSCFG_KEY_LOCATION) &
-			EFUSE_SYSCFG_KEY_LOCATION_MASK)
-		max_key_index = NUM_KEYS_IN_ROM;
-
-	if (key_index >= max_key_index)
-		return -1;
-
-	key_revoke = sp_sram_readl(SECURITY_EFUSE_KEYS_REVOKE_OFFSET);
-
-	/* if the key is not revoked, return 1 */
-	if ((key_revoke & BIT(key_index)) == 0)
-		return 1;
-
-	/* no valid key found */
-	return -1;
-}
-
-int security_get_key_hash(uint32_t *hash, uint8_t key_index)
-{
-	uint32_t valid_key_base, max_key_index = NUM_KEYS_IN_EFUSE;
-	int i;
-
-	/* check efuse for key location */
-	if ((arcboot_rom_global.config_fuse >> EFUSE_SYSCFG_KEY_LOCATION) &
-			EFUSE_SYSCFG_KEY_LOCATION_MASK) {
-
-		uint32_t *ptr;
-		max_key_index = NUM_KEYS_IN_ROM;
-
-		if (key_index >= max_key_index)
-			return -1;
-
-		/* read from ROM itself */
-		ptr = (uint32_t *)rom_key_hashes[key_index];
-		for (i = 0; i < SECURITY_DWORD_PER_KEY; i++)
-			*(hash + i) = ptr[i];
-	} else {
-
-		if (key_index >= max_key_index)
-			return -1;
-
-		/* read from efuse/SP SRAM */
-		valid_key_base = (key_index * SECURITY_BYTE_PER_DWORD *
-			  SECURITY_DWORD_PER_KEY) + SECURITY_EFUSE_KEYS_OFFSET;
-
-		for (i = 0; i < SECURITY_DWORD_PER_KEY; i++)
-			*(hash + i) = sp_sram_readl(valid_key_base + (i *
-					    SECURITY_BYTE_PER_DWORD));
-	}
-
-	return 1;
-}
-/*
- * arcboot - Signature Verification
- *
- * A simple application which call phase0b to validate the signature
- * of a blob of data.
- *
- * Both the protected data and the signature of the protected data can be
- * changed, but the public key should be locked into the device.
- *
- * test_phase0b_data     - blob of protected data
- * test_phase0b_data_len - size of the blob
- * test_phase0b_sig      - signature of the blob
- * test_phase0b_sig_len  - signature size
- * test_public_key       - public key
- * test_public_key_len   - public key size
- */
-
-int secureboot_verify_image(uint32_t image_size,
-				uint32_t image_offset)
-{
-	int verify = 0, ret, key_index;
-	unsigned char expected_hash[SHA384_SUM_LEN];
-	unsigned char actual_hash[SHA384_SUM_LEN];
-	uint32_t config_fuse;
-	uint32_t test_public_key_len = KEY_SIZE;
-	uint32_t test_sig_len = SIG_SIZE;
-	uint32_t test_data_len = image_size;
-	unsigned char *test_data = (unsigned char *)(image_offset);
-	unsigned char *test_sig = test_data + test_data_len;
-
-	/* Key header is located as the last 2 DWORDS of the image */
-	unsigned char *ptr_key_index = test_sig - KEY_PROP_HDR_SZ;
-        unsigned char *ptr_key_prop = test_sig - TOTAL_PUB_KEY_PROPS_SZ;
-        unsigned char *test_public_key =  ptr_key_prop - test_public_key_len;
-
-	uint32_t fw_version = base_readl((uint32_t)test_data);
-
-	/* check efuse for security validation enable */
-	config_fuse = sp_sram_readl(SECURITY_EFUSE_SYS_CFG_OFFSET);
-
-	if ((config_fuse & EFUSE_SYSCFG_ENABLE_VALIDATION_MASK) == 0)
-		return SECUREBOOT_VERIFY_SUCCESS;
-
-	/* check minimal valid security FW version */
-	if (security_check_valid_fw_version(fw_version) !=
-					SECURITY_FW_VERSION_SECURE)
-		return SECUREBOOT_VERIFY_VER_FAILED;
-
-	key_index = *(uint32_t *)ptr_key_index;
-
-	ret = security_check_key_revoke_status(key_index);
-	if (ret != 1)
-		return SECUREBOOT_VERIFY_KEY_FAILED;
-
-	ret = security_get_key_hash((uint32_t *)expected_hash, key_index);
-	if (ret != 1)
-		return SECUREBOOT_VERIFY_KEY_FAILED;
-
-	/* check pub key hash versus efuse */
-        sha384_csum(test_public_key, test_public_key_len, actual_hash, 0);
-
-        verify = memcmp(actual_hash, expected_hash, SHA384_SUM_LEN);
-	if (verify)
-		return SECUREBOOT_VERIFY_KEY_FAILED;
-
-	ret = rsa_verify(test_data, test_data_len, test_public_key,
-			test_public_key_len, test_sig, test_sig_len);
-
-	/* 0 - SUCCESS, but arcboot main code, expects 1 for SUCCESS */
-	if (!ret)
-		return SECUREBOOT_VERIFY_SUCCESS;
-
-	return SECUREBOOT_VERIFY_FAILED;
-}
diff --git a/gaudi2_arc/security.c b/gaudi2_arc/security.c
deleted file mode 100644
index 671db24..0000000
--- a/gaudi2_arc/security.c
+++ /dev/null
@@ -1,165 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
-#include "arcboot.h"
-#include "security.h"
-#include "efuse.h"
-#include "security.h"
-#include "spi_flash.h"
-
-static void security_disable_i2c_s_mstr_if(void)
-{
-	/* TODO */
-}
-
-static void security_disable_smi_mstr_if(void)
-{
-}
-
-void security_disable_jtag_mstr_if(void)
-{
-}
-
-void security_enable_jtag_mstr_if(void)
-{
-}
-
-void security_disable_twi_if(void)
-{
-	security_disable_i2c_s_mstr_if();
-	security_disable_smi_mstr_if();
-}
-
-void security_jtag_dap(__attribute__ ((unused)) uint8_t status)
-{
-	/* TODO: close dap WA jira: H5-1803 */
-}
-
-void security_debug_if_shared(void)
-{
-	/* Configure SMI,I2C and JTAG to be "shared" engine */
-}
-
-static void security_dcon_rr_lbw_config(
-			__attribute__ ((unused))uint8_t spi_controller_en)
-{
-}
-
-static void security_dcon_if0_rr_hbw_config(
-			__attribute__ ((unused))uint8_t spi_controller_en)
-{
-}
-
-static void security_dcon_if1_rr_hbw_config(
-			__attribute__ ((unused))uint8_t spi_controller_en)
-{
-}
-
-void security_i2c_s_rr_hbw_config(
-			__attribute__ ((unused))uint8_t spi_controller_en)
-{
-}
-
-void security_smi_rr_hbw_config(
-			__attribute__ ((unused))uint8_t spi_controller_en)
-{
-}
-
-void security_jtag_rr_hbw_config(
-			__attribute__ ((unused))uint8_t spi_controller_en)
-{
-}
-
-void security_pcie_rr_config(uint8_t spi_en)
-{
-	regmap_writel(PCIE_WRAP_PCIE_PROT_OVR_WR_EN_MASK |
-		      PCIE_WRAP_PCIE_PROT_OVR_RD_EN_MASK,
-		      mmPCIE_WRAP_PCIE_PROT_OVR);
-
-	/* NO Shared Initiator bits for Gaudi2, what to do? */
-	/* regmap_writel(PCIE_MSTR_RR_MSTR_IF_AXUSER_CORE_SHARED_MASK,
-		      mmPCIE_MSTR_RR_MSTR_IF_AXUSER_CORE);
-	*/
-	security_elbi_rr_lbw_config(spi_en);
-	security_dcon_rr_lbw_config(spi_en);
-	security_dcon_if0_rr_hbw_config(spi_en);
-	security_dcon_if1_rr_hbw_config(spi_en);
-}
-
-uint32_t security_key_soc_location_offset(uint32_t hash_key_index)
-{
-	uint32_t ret;
-	uint8_t key_soc_base_index;
-	uint8_t key_soc_location_index;
-	uint32_t key_soc_location[] = {
-		mmPSOC_SECURITY_PUBLIC_KEY_0_0,
-		mmPSOC_SECURITY_PUBLIC_KEY_1_0,
-		mmPSOC_SECURITY_PUBLIC_KEY_2_0,
-	};
-
-	/* select which index to choose */
-	key_soc_base_index = hash_key_index / SECURITY_DWORD_PER_KEY;
-	/* key is 384bit, 32bit * 12... which of the 12 location is requested */
-	key_soc_location_index = hash_key_index % SECURITY_DWORD_PER_KEY;
-
-	/* get base address of selected key index */
-	ret = key_soc_location[key_soc_base_index];
-	/* add to base address of key the spicifice location of currect DWORD */
-	ret += (key_soc_location_index * sizeof(uint32_t));
-
-	return ret;
-}
-
-void rom_get_hash(uint32_t *key_hash)
-{
-	uint32_t *ptr = (uint32_t *)rom_key_hashes;
-
-	for (int i = 0; i < SECURITY_KEY_HASH_ARRAY_SIZE; i++)
-		*(key_hash + i) = *(ptr + i);
-}
-
-void security_load_keys(void)
-{
-	uint8_t hash_key_index, num_dwords;
-	uint32_t key_hash[SECURITY_KEY_HASH_ARRAY_SIZE];
-	enum security_key_location key_location;
-
-	key_location = ((arcboot_rom_global.config_fuse >>
-			EFUSE_SYSCFG_KEY_LOCATION) &
-			EFUSE_SYSCFG_KEY_LOCATION_MASK);
-
-	arcboot_rom_status_update(PROGRESS_KEY_LOC);
-	arcboot_rom_status_update_location(PROGRESS_KEY_LOC_0, key_location);
-
-	switch (key_location) {
-	case KEY_LOCATION_EFUSE:
-		efuse_get_hash(key_hash);
-		break;
-	case KEY_LOCATION_ROM:
-	default:
-		rom_get_hash(key_hash);
-		break;
-	}
-
-	/* Store 3 key hashes in PSOC SECURITY, due to limited space */
-	num_dwords = NUM_KEYS_IN_EFUSE * SECURITY_DWORD_PER_KEY;
-	for (hash_key_index = 0; hash_key_index < num_dwords;
-	     hash_key_index++)
-		/* store habana key hash in PSOC_SECURITY */
-		regmap_writel(key_hash[hash_key_index],
-			      security_key_soc_location_offset(hash_key_index));
-
-	/* Store the 3 keys in SPSRAM */
-	for (hash_key_index = 0; hash_key_index < num_dwords;
-		hash_key_index++)
-		sp_sram_writel(key_hash[hash_key_index],
-			       SECURITY_EFUSE_KEYS_OFFSET + (hash_key_index *
-			       SECURITY_BYTE_PER_DWORD));
-}
-
-void check_and_enable_4b_termination(void)
-{
-	/* Not needed for Gaudi2 */
-}
diff --git a/gaudi2_arc/soc.h b/gaudi2_arc/soc.h
deleted file mode 100644
index 12ac0cb..0000000
--- a/gaudi2_arc/soc.h
+++ /dev/null
@@ -1,37 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2021 HabanaLabs Ltd.
- */
-
-#ifndef _SOC_H_
-#define _SOC_H_
-#include <gaudi2_regs.h>
-
-#define CPU_FREQ	500000000
-#define CPU_FREQ_US	500	/* CPU_FREQ / 1000000 */
-
-#define FW_SIZE					65535
-#define SIG_SIZE				384
-#define KEY_SIZE				384
-#define BLOB_START				4
-
-#define SPI_FLASH_BLOCK_SIZE			1024
-
-#define HL_SRAM_BASE				0x1000007FF0000000ULL
-#define HL_SP_SRAM_BASE            		0x1000007FF7FE0000ULL
-
-/*
- * This is not SDRAM Base, but the base address of General region
- * into which the rest of the components are mapped
- */
-#define HL_SDRAM_BASE				0x20000000
-#define HL_FLASH_OFFSET				0x6000000
-#define HL_ICCM_OFFSET				0x7040000
-#define HL_SBROM_OFFSET				0x7FC0000
-#define HL_FWSRAM_OFFSET			0x7FD0000
-#define HL_SCRATCHPAD_SRAM_OFFSET		0x7FE0000
-#define HL_PCIE_FWSRAM_OFFSET			0x7FF0000
-
-#define HL_CFG_BASE				0xF8000000
-
-#endif /* _SOC_H_ */
diff --git a/gaudi2_arc_gnu/Makefile b/gaudi2_arc_gnu/Makefile
new file mode 100644
index 0000000..1c6d192
--- /dev/null
+++ b/gaudi2_arc_gnu/Makefile
@@ -0,0 +1 @@
+LDFLAGS += -static
diff --git a/gaudi2_arc_gnu/arcboot_ram.lds b/gaudi2_arc_gnu/arcboot_ram.lds
new file mode 100644
index 0000000..056dca1
--- /dev/null
+++ b/gaudi2_arc_gnu/arcboot_ram.lds
@@ -0,0 +1,88 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	ICCM	: ORIGIN = 0x00000000, LENGTH = 128K
+	SRAM	: ORIGIN = 0x11000000, LENGTH = 48M
+	STM	: ORIGIN = 0x24000000, LENGTH = 32M
+	SPI	: ORIGIN = 0x26000000, LENGTH = 16M
+	SBROM	: ORIGIN = 0x27FC0000, LENGTH = 64K
+	FWSRAM	: ORIGIN = 0x27FD0000, LENGTH = 64K
+	CS	: ORIGIN = 0x2C000000, LENGTH = 32M
+	DCCM	: ORIGIN = 0x80000000, LENGTH = 32K
+}
+
+REGION_ALIAS("REGION_TEXT",	ICCM)
+REGION_ALIAS("REGION_DATA",	DCCM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_SBSS",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_HEAP",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+ENTRY(__start)
+
+SECTIONS
+{
+	.text : {
+		app/crt0_ram.o (.text*)
+		*(.text .text.*)
+		_etext = . ;
+	} > REGION_TEXT
+
+	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
+		_data = .;
+		*(.data)
+	} > REGION_DATA
+
+	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
+		__SDATA_BEGIN__ = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_edata = .;
+		PROVIDE (edata = .);
+	} > REGION_SDATA
+
+	.sbss : {
+		__bss_start = .;
+		PROVIDE (__sbss_start = .);
+		PROVIDE (___sbss_start = .);
+		*(.dynsbss)
+		*(.sbss .sbss.* .gnu.linkonce.sb.*)
+		*(.scommon)
+		PROVIDE (__sbss_end = .);
+		PROVIDE (___sbss_end = .);
+	} > REGION_SBSS
+
+	.bss : {
+		*(.dynbss)
+		*(.bss .bss.* .gnu.linkonce.b.*)
+		*(COMMON)
+		/* Align here to ensure that the .bss section occupies space up to
+		_end. Align after .bss to ensure correct alignment even if the
+		.bss section disappears because there are no input sections. */
+		. = ALIGN(32 / 8);
+		_end = .;
+		PROVIDE (end = .);
+	} > REGION_BSS
+
+	.heap : {
+		. = ALIGN(32 / 8);
+		__start_heap = . ;
+		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 1k)) ;
+		__end_heap = . ;
+		. = ALIGN(0x8);
+	} > REGION_HEAP
+
+	.stack : {
+		. = ALIGN(32 / 8);
+		__stack = . ;
+		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 512)) ;
+		__stack_top = . ;
+	} > REGION_STACK
+}
diff --git a/gaudi2_arc_gnu/arcboot_rom.lds b/gaudi2_arc_gnu/arcboot_rom.lds
new file mode 100644
index 0000000..1721eb2
--- /dev/null
+++ b/gaudi2_arc_gnu/arcboot_rom.lds
@@ -0,0 +1,96 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	ICCM	: ORIGIN = 0x00000000, LENGTH = 128K
+	SRAM	: ORIGIN = 0x11000000, LENGTH = 48M
+	STM	: ORIGIN = 0x24000000, LENGTH = 32M
+	SPI	: ORIGIN = 0x26000000, LENGTH = 16M
+	SBROM	: ORIGIN = 0x27FC0000, LENGTH = 64K
+	FWSRAM	: ORIGIN = 0x27FD0000, LENGTH = 64K
+	CS	: ORIGIN = 0x2C000000, LENGTH = 32M
+	DCCM	: ORIGIN = 0x80000000, LENGTH = 32K
+}
+
+REGION_ALIAS("REGION_IVT",	SBROM)
+REGION_ALIAS("REGION_TEXT",	SBROM)
+REGION_ALIAS("REGION_DATA",	DCCM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_SBSS",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_HEAP",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+ENTRY(__start)
+
+SECTIONS
+{
+	_dccm_begin = ORIGIN(DCCM);
+	_dccm_end = ORIGIN(DCCM) + LENGTH(DCCM);
+
+	.ivt : {
+		KEEP (*(.ivt));
+	} > REGION_IVT
+
+	.text : {
+		app/crt0_rom.o (.text*)
+		*(.text .text.*)
+		_etext = . ;
+	} > REGION_TEXT
+
+	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
+		_data = .;
+		*(.data)
+	} > REGION_DATA
+
+	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
+		__SDATA_BEGIN__ = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_edata = .;
+		PROVIDE (edata = .);
+	} > REGION_SDATA
+
+	.sbss : {
+		__bss_start = .;
+		PROVIDE (__sbss_start = .);
+		PROVIDE (___sbss_start = .);
+		*(.dynsbss)
+		*(.sbss .sbss.* .gnu.linkonce.sb.*)
+		*(.scommon)
+		PROVIDE (__sbss_end = .);
+		PROVIDE (___sbss_end = .);
+	} > REGION_SBSS
+
+	.bss : {
+		*(.dynbss)
+		*(.bss .bss.* .gnu.linkonce.b.*)
+		*(COMMON)
+		/* Align here to ensure that the .bss section occupies space up to
+		_end. Align after .bss to ensure correct alignment even if the
+		.bss section disappears because there are no input sections. */
+		. = ALIGN(32 / 8);
+		_end = .;
+		PROVIDE (end = .);
+	} > REGION_BSS
+
+	.heap : {
+		. = ALIGN(32 / 8);
+		__start_heap = . ;
+		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 5k)) ;
+		__end_heap = . ;
+		. = ALIGN(0x8);
+	} > REGION_HEAP
+
+	.stack : {
+		. = ALIGN(32 / 8);
+		__stack = . ;
+		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 2k)) ;
+		__stack_top = . ;
+	} > REGION_STACK
+}
diff --git a/gaudi2_arc_gnu/asic.c b/gaudi2_arc_gnu/asic.c
new file mode 100644
index 0000000..7b6e440
--- /dev/null
+++ b/gaudi2_arc_gnu/asic.c
@@ -0,0 +1,4 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
diff --git a/gaudi2_arc_gnu/asic_arcboot_ram.lds b/gaudi2_arc_gnu/asic_arcboot_ram.lds
new file mode 100644
index 0000000..827fe1f
--- /dev/null
+++ b/gaudi2_arc_gnu/asic_arcboot_ram.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc_gnu/asic_arcboot_rom.lds b/gaudi2_arc_gnu/asic_arcboot_rom.lds
new file mode 100644
index 0000000..9e1e858
--- /dev/null
+++ b/gaudi2_arc_gnu/asic_arcboot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_rom.lds
diff --git a/gaudi2_arc_gnu/common.c b/gaudi2_arc_gnu/common.c
new file mode 100644
index 0000000..af86bf7
--- /dev/null
+++ b/gaudi2_arc_gnu/common.c
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <unistd.h>
+#include <efuse.h>
+#include <pcie.h>
+#include <arcboot.h>
+#include <psoc_efuse_regs.h>
+#include <psoc_global_conf_masks.h>
+#include <psoc_reset_conf_masks.h>
+#include <psoc_reset_conf_regs.h>
+
+#define ARCBOOT_STATUS_REG_SZ_BITS	32
+
+void arcboot_rom_status_update(enum arcboot_rom_progress progress)
+{
+	uint32_t reg, offset;
+
+	offset = progress / ARCBOOT_STATUS_REG_SZ_BITS;
+	progress = progress % ARCBOOT_STATUS_REG_SZ_BITS;
+
+	reg = regmap_readl((mmPSOC_SECURITY_SW_STATUS_0 + offset * 4));
+	reg |= BIT(progress);
+	regmap_writel(reg, (mmPSOC_SECURITY_SW_STATUS_0 + offset  * 4));
+}
+
+uint32_t get_efuse_read_reg(void)
+{
+	return mmPSOC_EFUSE_READ_CTRL;
+}
+
+uint32_t get_mem_cfg_mask(void)
+{
+	return PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_MEM_SIZE_MASK;
+}
+
+uint32_t get_fld_size_shift(void)
+{
+	return PSOC_GLOBAL_CONF_SPI_DMA_CTRL_MEM_SIZE_SHIFT;
+}
+
+void update_warm_reset_config(void)
+{
+	regmap_writel(PSOC_RESET_CONF_PSOC_PRSTN_RST_CFG_EN_MASK,
+			mmPSOC_RESET_CONF_PSOC_PRSTN_RST_CFG);
+	regmap_writel(PSOC_RESET_CONF_ARC_PRSTN_RST_CFG_EN_MASK,
+			mmPSOC_RESET_CONF_ARC_PRSTN_RST_CFG);
+}
+
+uint8_t get_sris_mode_from_strap()
+{
+	uint32_t boot_strap;
+
+	boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H);
+
+	return !!(boot_strap & PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H_SRIS_MODE_MASK);
+}
+
+void enable_flr_handling_fsm(void)
+{
+	/* TODO: No HW FSM bits in Gaudi2, wait for ORON to comeback */
+}
diff --git a/gaudi2_arc_gnu/dev_security.h b/gaudi2_arc_gnu/dev_security.h
new file mode 100644
index 0000000..28a618e
--- /dev/null
+++ b/gaudi2_arc_gnu/dev_security.h
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef _GAUDI2_DEV_SECURITY_H_
+#define _GAUDI2_DEV_SECURITY_H_
+
+#include <sha512.h>
+
+#define NUM_KEYS_IN_ROM					5
+#define NUM_KEYS_IN_EFUSE				3
+#define SECURITY_DWORD_PER_KEY				12
+#define SECURITY_KEY_HASH_COUNT				3
+
+enum security_key_location {
+	KEY_LOCATION_EFUSE,
+	KEY_LOCATION_ROM
+};
+
+/* size of key properties in bytes */
+#define KEY_PROP_RR_SZ  				384
+#define KEY_PROP_MOD_SZ					384
+#define KEY_PROP_EXP_SZ					8
+#define KEY_PROP_INV_SZ					4
+#define KEY_PROP_NB_SZ					4
+#define KEY_PROP_HDR_SZ					8
+
+/* 384 + 384 + 8 + 4 + 4 + 8 = 792 */
+#define TOTAL_PUB_KEY_PROPS_SZ  (KEY_PROP_RR_SZ + KEY_PROP_MOD_SZ + \
+				KEY_PROP_EXP_SZ + KEY_PROP_INV_SZ + \
+				KEY_PROP_NB_SZ + KEY_PROP_HDR_SZ)
+
+extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
+
+#endif /* _GAUDI2_DEV_SECURITY_H_ */
diff --git a/gaudi2_arc_gnu/onix.c b/gaudi2_arc_gnu/onix.c
new file mode 100644
index 0000000..f82fc69
--- /dev/null
+++ b/gaudi2_arc_gnu/onix.c
@@ -0,0 +1,29 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+#include <unistd.h>
+#include "arcboot.h"
+#include "efuse.h"
+#include <common.h>
+#include <psoc_global_conf_masks.h>
+
+#define EFUSE_SYSCFG_ONIX_DEFAULT \
+				(EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI | \
+				 EFUSE_SYSCFG_KEY_LOCATION_SPI | \
+				 EFUSE_SYSCFG_REVOKATION_LOCATION_SPI | \
+				 EFUSE_SYSCFG_ENABLE_VALIDATION_MASK | \
+				 EFUSE_SYSCFG_GPIO02_RECOVERY_CONTROL)
+
+int fuse_read(uint32_t word, uint32_t *val)
+{
+	int ret = -1;
+
+	if (word == EFUSE_SYSCFG_OFFSET) {
+		*val = EFUSE_SYSCFG_ONIX_DEFAULT;
+		ret = 0;
+	}
+
+	return ret;
+}
diff --git a/gaudi2_arc_gnu/onix_arcboot_ram.lds b/gaudi2_arc_gnu/onix_arcboot_ram.lds
new file mode 100644
index 0000000..827fe1f
--- /dev/null
+++ b/gaudi2_arc_gnu/onix_arcboot_ram.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc_gnu/onix_arcboot_rom.lds b/gaudi2_arc_gnu/onix_arcboot_rom.lds
new file mode 100644
index 0000000..1a9a352
--- /dev/null
+++ b/gaudi2_arc_gnu/onix_arcboot_rom.lds
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+INCLUDE arcboot_rom.lds
+
+/* This is SRAM address, used for Onix targets */
+MEMORY
+{
+	SBROM	: ORIGIN = 0x11000000, LENGTH = 64K
+}
diff --git a/gaudi2_arc_gnu/pldm.c b/gaudi2_arc_gnu/pldm.c
new file mode 100644
index 0000000..7b6e440
--- /dev/null
+++ b/gaudi2_arc_gnu/pldm.c
@@ -0,0 +1,4 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
diff --git a/gaudi2_arc_gnu/pldm_arcboot_ram.lds b/gaudi2_arc_gnu/pldm_arcboot_ram.lds
new file mode 100644
index 0000000..827fe1f
--- /dev/null
+++ b/gaudi2_arc_gnu/pldm_arcboot_ram.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc_gnu/pldm_arcboot_rom.lds b/gaudi2_arc_gnu/pldm_arcboot_rom.lds
new file mode 100644
index 0000000..9e1e858
--- /dev/null
+++ b/gaudi2_arc_gnu/pldm_arcboot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_rom.lds
diff --git a/gaudi2_arc_gnu/secureboot.c b/gaudi2_arc_gnu/secureboot.c
new file mode 100644
index 0000000..2ef2f58
--- /dev/null
+++ b/gaudi2_arc_gnu/secureboot.c
@@ -0,0 +1,196 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+#include <string.h>
+#include <arcboot.h>
+#include <efuse.h>
+#include <secureboot.h>
+#include <security.h>
+#include <dev_security.h>
+#include <rsa.h>
+#include <sha512.h>
+
+
+const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN] = {
+	/* Habana Public Key */
+	{
+		0x09, 0xa8, 0xdd, 0xac, 0xbe, 0x8a, 0x6d, 0x34,
+		0x94, 0x23, 0x62, 0x7c, 0xd6, 0x11, 0xe8, 0x15,
+		0xb4, 0x44, 0xe2, 0x8d, 0x81, 0xc9, 0xc5, 0xd0,
+		0x49, 0x3d, 0xd6, 0x4f, 0x10, 0xac, 0xdf, 0x83,
+		0x04, 0x61, 0xb5, 0xde, 0x3e, 0x58, 0x5a, 0x73,
+		0x18, 0xb4, 0x56, 0xbc, 0xcc, 0x91, 0xf0, 0xed
+	},
+	/* Additional Keys - TODO */
+	{
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0
+	},
+	{
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0
+	},
+	{
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0
+	},
+	{
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0
+	}
+};
+
+static int security_check_key_revoke_status(uint32_t key_index)
+{
+	uint8_t max_key_index = NUM_KEYS_IN_EFUSE, key_revoke;
+
+	/*
+	 * Key index ranges from:
+	 * 0 - 4: if key location == ROM (1)
+	 * 0 - 2: if key location == eFUSE (0)
+	 */
+	/* check efuse for key location */
+	if ((arcboot_rom_global.config_fuse >> EFUSE_SYSCFG_KEY_LOCATION) &
+			EFUSE_SYSCFG_KEY_LOCATION_MASK)
+		max_key_index = NUM_KEYS_IN_ROM;
+
+	if (key_index >= max_key_index)
+		return -1;
+
+	key_revoke = sp_sram_readl(SECURITY_EFUSE_KEYS_REVOKE_OFFSET);
+
+	/* if the key is not revoked, return 1 */
+	if ((key_revoke & BIT(key_index)) == 0)
+		return 1;
+
+	/* no valid key found */
+	return -1;
+}
+
+int security_get_key_hash(uint32_t *hash, uint8_t key_index)
+{
+	uint32_t valid_key_base, max_key_index = NUM_KEYS_IN_EFUSE;
+	int i;
+
+	/* check efuse for key location */
+	if ((arcboot_rom_global.config_fuse >> EFUSE_SYSCFG_KEY_LOCATION) &
+			EFUSE_SYSCFG_KEY_LOCATION_MASK) {
+
+		uint32_t *ptr;
+		max_key_index = NUM_KEYS_IN_ROM;
+
+		if (key_index >= max_key_index)
+			return -1;
+
+		/* read from ROM itself */
+		ptr = (uint32_t *)rom_key_hashes[key_index];
+		for (i = 0; i < SECURITY_DWORD_PER_KEY; i++)
+			*(hash + i) = ptr[i];
+	} else {
+
+		if (key_index >= max_key_index)
+			return -1;
+
+		/* read from efuse/SP SRAM */
+		valid_key_base = (key_index * SECURITY_BYTE_PER_DWORD *
+			  SECURITY_DWORD_PER_KEY) + SECURITY_EFUSE_KEYS_OFFSET;
+
+		for (i = 0; i < SECURITY_DWORD_PER_KEY; i++)
+			*(hash + i) = sp_sram_readl(valid_key_base + (i *
+					    SECURITY_BYTE_PER_DWORD));
+	}
+
+	return 1;
+}
+/*
+ * arcboot - Signature Verification
+ *
+ * A simple application which call phase0b to validate the signature
+ * of a blob of data.
+ *
+ * Both the protected data and the signature of the protected data can be
+ * changed, but the public key should be locked into the device.
+ *
+ * test_phase0b_data     - blob of protected data
+ * test_phase0b_data_len - size of the blob
+ * test_phase0b_sig      - signature of the blob
+ * test_phase0b_sig_len  - signature size
+ * test_public_key       - public key
+ * test_public_key_len   - public key size
+ */
+
+int secureboot_verify_image(uint32_t image_size,
+				uint32_t image_offset)
+{
+	int verify = 0, ret, key_index;
+	unsigned char expected_hash[SHA384_SUM_LEN];
+	unsigned char actual_hash[SHA384_SUM_LEN];
+	uint32_t config_fuse;
+	uint32_t test_public_key_len = KEY_SIZE;
+	uint32_t test_sig_len = SIG_SIZE;
+	uint32_t test_data_len = image_size;
+	unsigned char *test_data = (unsigned char *)(image_offset);
+	unsigned char *test_sig = test_data + test_data_len;
+
+	/* Key header is located as the last 2 DWORDS of the image */
+	unsigned char *ptr_key_index = test_sig - KEY_PROP_HDR_SZ;
+        unsigned char *ptr_key_prop = test_sig - TOTAL_PUB_KEY_PROPS_SZ;
+        unsigned char *test_public_key =  ptr_key_prop - test_public_key_len;
+
+	uint32_t fw_version = base_readl((uint32_t)test_data);
+
+	/* check efuse for security validation enable */
+	config_fuse = sp_sram_readl(SECURITY_EFUSE_SYS_CFG_OFFSET);
+
+	if ((config_fuse & EFUSE_SYSCFG_ENABLE_VALIDATION_MASK) == 0)
+		return SECUREBOOT_VERIFY_SUCCESS;
+
+	/* check minimal valid security FW version */
+	if (security_check_valid_fw_version(fw_version) !=
+					SECURITY_FW_VERSION_SECURE)
+		return SECUREBOOT_VERIFY_VER_FAILED;
+
+	key_index = *(uint32_t *)ptr_key_index;
+
+	ret = security_check_key_revoke_status(key_index);
+	if (ret != 1)
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	ret = security_get_key_hash((uint32_t *)expected_hash, key_index);
+	if (ret != 1)
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	/* check pub key hash versus efuse */
+        sha384_csum(test_public_key, test_public_key_len, actual_hash, 0);
+
+        verify = memcmp(actual_hash, expected_hash, SHA384_SUM_LEN);
+	if (verify)
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	ret = rsa_verify(test_data, test_data_len, test_public_key,
+			test_public_key_len, test_sig, test_sig_len);
+
+	/* 0 - SUCCESS, but arcboot main code, expects 1 for SUCCESS */
+	if (!ret)
+		return SECUREBOOT_VERIFY_SUCCESS;
+
+	return SECUREBOOT_VERIFY_FAILED;
+}
diff --git a/gaudi2_arc_gnu/security.c b/gaudi2_arc_gnu/security.c
new file mode 100644
index 0000000..671db24
--- /dev/null
+++ b/gaudi2_arc_gnu/security.c
@@ -0,0 +1,165 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+#include "arcboot.h"
+#include "security.h"
+#include "efuse.h"
+#include "security.h"
+#include "spi_flash.h"
+
+static void security_disable_i2c_s_mstr_if(void)
+{
+	/* TODO */
+}
+
+static void security_disable_smi_mstr_if(void)
+{
+}
+
+void security_disable_jtag_mstr_if(void)
+{
+}
+
+void security_enable_jtag_mstr_if(void)
+{
+}
+
+void security_disable_twi_if(void)
+{
+	security_disable_i2c_s_mstr_if();
+	security_disable_smi_mstr_if();
+}
+
+void security_jtag_dap(__attribute__ ((unused)) uint8_t status)
+{
+	/* TODO: close dap WA jira: H5-1803 */
+}
+
+void security_debug_if_shared(void)
+{
+	/* Configure SMI,I2C and JTAG to be "shared" engine */
+}
+
+static void security_dcon_rr_lbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+static void security_dcon_if0_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+static void security_dcon_if1_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+void security_i2c_s_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+void security_smi_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+void security_jtag_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+void security_pcie_rr_config(uint8_t spi_en)
+{
+	regmap_writel(PCIE_WRAP_PCIE_PROT_OVR_WR_EN_MASK |
+		      PCIE_WRAP_PCIE_PROT_OVR_RD_EN_MASK,
+		      mmPCIE_WRAP_PCIE_PROT_OVR);
+
+	/* NO Shared Initiator bits for Gaudi2, what to do? */
+	/* regmap_writel(PCIE_MSTR_RR_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+		      mmPCIE_MSTR_RR_MSTR_IF_AXUSER_CORE);
+	*/
+	security_elbi_rr_lbw_config(spi_en);
+	security_dcon_rr_lbw_config(spi_en);
+	security_dcon_if0_rr_hbw_config(spi_en);
+	security_dcon_if1_rr_hbw_config(spi_en);
+}
+
+uint32_t security_key_soc_location_offset(uint32_t hash_key_index)
+{
+	uint32_t ret;
+	uint8_t key_soc_base_index;
+	uint8_t key_soc_location_index;
+	uint32_t key_soc_location[] = {
+		mmPSOC_SECURITY_PUBLIC_KEY_0_0,
+		mmPSOC_SECURITY_PUBLIC_KEY_1_0,
+		mmPSOC_SECURITY_PUBLIC_KEY_2_0,
+	};
+
+	/* select which index to choose */
+	key_soc_base_index = hash_key_index / SECURITY_DWORD_PER_KEY;
+	/* key is 384bit, 32bit * 12... which of the 12 location is requested */
+	key_soc_location_index = hash_key_index % SECURITY_DWORD_PER_KEY;
+
+	/* get base address of selected key index */
+	ret = key_soc_location[key_soc_base_index];
+	/* add to base address of key the spicifice location of currect DWORD */
+	ret += (key_soc_location_index * sizeof(uint32_t));
+
+	return ret;
+}
+
+void rom_get_hash(uint32_t *key_hash)
+{
+	uint32_t *ptr = (uint32_t *)rom_key_hashes;
+
+	for (int i = 0; i < SECURITY_KEY_HASH_ARRAY_SIZE; i++)
+		*(key_hash + i) = *(ptr + i);
+}
+
+void security_load_keys(void)
+{
+	uint8_t hash_key_index, num_dwords;
+	uint32_t key_hash[SECURITY_KEY_HASH_ARRAY_SIZE];
+	enum security_key_location key_location;
+
+	key_location = ((arcboot_rom_global.config_fuse >>
+			EFUSE_SYSCFG_KEY_LOCATION) &
+			EFUSE_SYSCFG_KEY_LOCATION_MASK);
+
+	arcboot_rom_status_update(PROGRESS_KEY_LOC);
+	arcboot_rom_status_update_location(PROGRESS_KEY_LOC_0, key_location);
+
+	switch (key_location) {
+	case KEY_LOCATION_EFUSE:
+		efuse_get_hash(key_hash);
+		break;
+	case KEY_LOCATION_ROM:
+	default:
+		rom_get_hash(key_hash);
+		break;
+	}
+
+	/* Store 3 key hashes in PSOC SECURITY, due to limited space */
+	num_dwords = NUM_KEYS_IN_EFUSE * SECURITY_DWORD_PER_KEY;
+	for (hash_key_index = 0; hash_key_index < num_dwords;
+	     hash_key_index++)
+		/* store habana key hash in PSOC_SECURITY */
+		regmap_writel(key_hash[hash_key_index],
+			      security_key_soc_location_offset(hash_key_index));
+
+	/* Store the 3 keys in SPSRAM */
+	for (hash_key_index = 0; hash_key_index < num_dwords;
+		hash_key_index++)
+		sp_sram_writel(key_hash[hash_key_index],
+			       SECURITY_EFUSE_KEYS_OFFSET + (hash_key_index *
+			       SECURITY_BYTE_PER_DWORD));
+}
+
+void check_and_enable_4b_termination(void)
+{
+	/* Not needed for Gaudi2 */
+}
diff --git a/gaudi2_arc_gnu/soc.h b/gaudi2_arc_gnu/soc.h
new file mode 100644
index 0000000..12ac0cb
--- /dev/null
+++ b/gaudi2_arc_gnu/soc.h
@@ -0,0 +1,37 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+#ifndef _SOC_H_
+#define _SOC_H_
+#include <gaudi2_regs.h>
+
+#define CPU_FREQ	500000000
+#define CPU_FREQ_US	500	/* CPU_FREQ / 1000000 */
+
+#define FW_SIZE					65535
+#define SIG_SIZE				384
+#define KEY_SIZE				384
+#define BLOB_START				4
+
+#define SPI_FLASH_BLOCK_SIZE			1024
+
+#define HL_SRAM_BASE				0x1000007FF0000000ULL
+#define HL_SP_SRAM_BASE            		0x1000007FF7FE0000ULL
+
+/*
+ * This is not SDRAM Base, but the base address of General region
+ * into which the rest of the components are mapped
+ */
+#define HL_SDRAM_BASE				0x20000000
+#define HL_FLASH_OFFSET				0x6000000
+#define HL_ICCM_OFFSET				0x7040000
+#define HL_SBROM_OFFSET				0x7FC0000
+#define HL_FWSRAM_OFFSET			0x7FD0000
+#define HL_SCRATCHPAD_SRAM_OFFSET		0x7FE0000
+#define HL_PCIE_FWSRAM_OFFSET			0x7FF0000
+
+#define HL_CFG_BASE				0xF8000000
+
+#endif /* _SOC_H_ */
